// igniteui-react-core/esm2015/lib/BollingerBandWidthIndicatorDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { StrategyBasedIndicatorDescription } from "./StrategyBasedIndicatorDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let BollingerBandWidthIndicatorDescription = /*@__PURE__*/ (() => {
    class BollingerBandWidthIndicatorDescription extends StrategyBasedIndicatorDescription {
        constructor() {
            super();
            this.eb = 0;
            this.d9 = 0;
        }
        get_type() {
            return "BollingerBandWidthIndicator";
        }
        get period() {
            return this.eb;
        }
        set period(a) {
            this.eb = a;
            this.e("Period");
        }
        get multiplier() {
            return this.d9;
        }
        set multiplier(a) {
            this.d9 = a;
            this.e("Multiplier");
        }
    }
    BollingerBandWidthIndicatorDescription.$t = /*@__PURE__*/ markType(BollingerBandWidthIndicatorDescription, 'BollingerBandWidthIndicatorDescription', StrategyBasedIndicatorDescription.$);
    return BollingerBandWidthIndicatorDescription;
})();
