// igniteui-react-core/esm2015/lib/ItemLegendDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { LegendBaseDescription } from "./LegendBaseDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let ItemLegendDescription = /*@__PURE__*/ (() => {
    class ItemLegendDescription extends LegendBaseDescription {
        constructor() {
            super();
            this.u = null;
            this.s = null;
            this.t = null;
        }
        get width() {
            return this.u;
        }
        set width(a) {
            this.u = a;
            this.e("Width");
        }
        get height() {
            return this.s;
        }
        set height(a) {
            this.s = a;
            this.e("Height");
        }
        get_type() {
            return "ItemLegend";
        }
        get orientation() {
            return this.t;
        }
        set orientation(a) {
            this.t = a;
            this.e("Orientation");
        }
    }
    ItemLegendDescription.$t = /*@__PURE__*/ markType(ItemLegendDescription, 'ItemLegendDescription', LegendBaseDescription.$);
    return ItemLegendDescription;
})();
