// igniteui-react-core/esm2015/lib/NativeUIDropDown.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { NativeUIDisplayDensity_$type } from "./NativeUIDisplayDensity";
import { enumGetBox, EnumUtil, markType } from "./type";
/**
 * @hidden
 */
export let NativeUIDropDown = /*@__PURE__*/ (() => {
    class NativeUIDropDown extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.aq = null;
        }
        get_c() {
            return 4;
        }
        get ap() {
            let a = this.o(1);
            return a == null ? 1 : EnumUtil.getEnumValue(NativeUIDisplayDensity_$type, a);
        }
        set ap(a) {
            this.ad(1, enumGetBox(NativeUIDisplayDensity_$type, a));
        }
        get ar() {
            return this.aq;
        }
        set ar(a) {
            let b = this.aq;
            this.aq = a;
            if (this.aq != null) {
                this.w(3, this.aq);
            }
            else {
                this.ab(3, b);
            }
        }
        get at() {
            let a = this.o(7);
            return a;
        }
        set at(a) {
            this.ad(7, a);
        }
        get as() {
            let a = this.o(6);
            return a;
        }
        set as(a) {
            this.ad(6, a);
        }
    }
    NativeUIDropDown.$t = /*@__PURE__*/ markType(NativeUIDropDown, 'NativeUIDropDown', NativeUIComponent.$);
    return NativeUIDropDown;
})();
