// igniteui-react-core/esm2015/lib/NativeUISlider.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { NativeUIComponent } from "./NativeUIComponent";
import { markType } from "./type";
/**
 * @hidden
 */
export let NativeUISlider = /*@__PURE__*/ (() => {
    class NativeUISlider extends NativeUIComponent {
        constructor() {
            super(...arguments);
            this.ap = null;
            this.aq = null;
        }
        get_c() {
            return 5;
        }
        get aw() {
            let a = this.o(2);
            return a;
        }
        set aw(a) {
            this.ad(2, a);
        }
        get au() {
            let a = this.o(10);
            return a;
        }
        set au(a) {
            this.ad(10, a);
        }
        get at() {
            let a = this.o(11);
            return a;
        }
        set at(a) {
            this.ad(11, a);
        }
        get av() {
            let a = this.o(12);
            return a;
        }
        set av(a) {
            this.ad(12, a);
        }
        get ar() {
            return this.ap;
        }
        set ar(a) {
            let b = this.ap;
            this.ap = a;
            if (this.ap != null) {
                this.w(4, this.ap);
            }
            else {
                this.ab(4, b);
            }
        }
        get as() {
            return this.aq;
        }
        set as(a) {
            let b = this.aq;
            this.aq = a;
            if (this.aq != null) {
                this.w(0, this.aq);
            }
            else {
                this.ab(0, b);
            }
        }
    }
    NativeUISlider.$t = /*@__PURE__*/ markType(NativeUISlider, 'NativeUISlider', NativeUIComponent.$);
    return NativeUISlider;
})();
