// igniteui-react-core/esm2015/lib/RadialColumnSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { AnchoredRadialSeriesDescription } from "./AnchoredRadialSeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RadialColumnSeriesDescription = /*@__PURE__*/ (() => {
    class RadialColumnSeriesDescription extends AnchoredRadialSeriesDescription {
        constructor() {
            super();
            this.ej = 0;
            this.ek = 0;
        }
        get_type() {
            return "RadialColumnSeries";
        }
        get radiusX() {
            return this.ej;
        }
        set radiusX(a) {
            this.ej = a;
            this.e("RadiusX");
        }
        get radiusY() {
            return this.ek;
        }
        set radiusY(a) {
            this.ek = a;
            this.e("RadiusY");
        }
    }
    RadialColumnSeriesDescription.$t = /*@__PURE__*/ markType(RadialColumnSeriesDescription, 'RadialColumnSeriesDescription', AnchoredRadialSeriesDescription.$);
    return RadialColumnSeriesDescription;
})();
