// igniteui-react-core/esm2015/lib/RangeColumnSeriesDescription.js
/*
THIS INFRAGISTICS ULTIMATE SOFTWARE LICENSE  AGREEMENT ("AGREEMENT") LOCATED HERE:
https://www.infragistics.com/legal/license/igultimate-la
https://www.infragistics.com/legal/license/igultimate-eula
GOVERNS THE LICENSING, INSTALLATION AND USE OF INFRAGISTICS SOFTWARE. BY DOWNLOADING AND/OR INSTALLING AND USING INFRAGISTICS SOFTWARE:  you are indicating that you have read and understand this Agreement, and agree to be legally bound by it on behalf of the yourself and your company.
*/
import { HorizontalRangeCategorySeriesDescription } from "./HorizontalRangeCategorySeriesDescription";
import { markType } from "./type";
/**
 * @hidden
 */
export let RangeColumnSeriesDescription = /*@__PURE__*/ (() => {
    class RangeColumnSeriesDescription extends HorizontalRangeCategorySeriesDescription {
        constructor() {
            super();
            this.eh = 0;
            this.ei = 0;
        }
        get_type() {
            return "RangeColumnSeries";
        }
        get radiusX() {
            return this.eh;
        }
        set radiusX(a) {
            this.eh = a;
            this.e("RadiusX");
        }
        get radiusY() {
            return this.ei;
        }
        set radiusY(a) {
            this.ei = a;
            this.e("RadiusY");
        }
    }
    RangeColumnSeriesDescription.$t = /*@__PURE__*/ markType(RangeColumnSeriesDescription, 'RangeColumnSeriesDescription', HorizontalRangeCategorySeriesDescription.$);
    return RangeColumnSeriesDescription;
})();
