// igniteui-react-charts/esm2015/lib/igr-radial-column-series.js
import { IgrAnchoredRadialSeries } from "./igr-anchored-radial-series";
import { RadialColumnSeries } from "./RadialColumnSeries";
/**
 * Represents a IgxDataChartComponent radial column series.
 * Compare values across categories by using radial rectangles.
*/
export class IgrRadialColumnSeries extends IgrAnchoredRadialSeries {
    createImplementation() {
        return new RadialColumnSeries();
    }
    /**
                                 * @hidden
                                 */
    get i() {
        return this._implementation;
    }
    constructor(props) {
        super(props);
    }
    /**
     * Gets whether the current series shows a column shape.
    */
    get isColumn() {
        return this.i.dh;
    }
    /**
     * Overridden by derived series classes to indicate when marker-less display is preferred or not.
    */
    get isMarkerlessDisplayPreferred() {
        return this.i.d2;
    }
    /**
     * Gets or sets the x-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusX() {
        return this.i.t7;
    }
    set radiusX(v) {
        this.i.t7 = +v;
    }
    /**
     * Gets or sets the y-radius of the ellipse that is used to round the corners of the column.
    */
    get radiusY() {
        return this.i.t8;
    }
    set radiusY(v) {
        this.i.t8 = +v;
    }
    /**
     * When overridden in a derived class, is invoked whenever application code or internal processes
     * call ApplyTemplate.
    
    */
    onApplyTemplate() {
        this.i.ac();
    }
    /**
     * For a category plotted series, returns the current width of the items within the categories. This only returns a value if the items have some form of width (e.g. columns, bars, etc.) otherwise 0 is returned.
    
    */
    getItemSpan() {
        let iv = this.i.ga();
        return (iv);
    }
}
